slicerator/__init__.py

```diff
@@ -389,39 +402,68 @@ def __init__(self, ancestor, proc_func, propagate_attrs=None):
         if propagate_attrs is not None:
-            self._propagate_attrs = propagate_attrs
+            self._propagate_attrs = set(propagate_attrs)
         else:
             # check propagated_attrs field from the ancestor definition
-            self._propagate_attrs = []
-            if hasattr(ancestor, '_propagate_attrs'):
-                self._propagate_attrs += ancestor._propagate_attrs
-            if hasattr(ancestor, 'propagate_attrs'):
-                self._propagate_attrs += ancestor.propagate_attrs
-
-            # add methods having the _propagate flag
-            for attr in _iter_attr(ancestor):
-                if hasattr(attr, '_propagate_flag'):
-                    self._propagate_attrs.append(attr.__name__)
-
-        self._ancestor = ancestor
-        self._proc_func = proc_func
+            self._propagate_attrs = set()
+            for a in self._get_prop_ancestors():
+                if hasattr(a, '_propagate_attrs'):
+                    self._propagate_attrs.update(a._propagate_attrs)
+                if hasattr(a, 'propagate_attrs'):
+                    self._propagate_attrs.update(a.propagate_attrs)
+
+                # add methods having the _propagate flag
+                for attr in _iter_attr(a):
+                    if hasattr(attr, '_propagate_flag'):
+                        self._propagate_attrs.add(attr.__name__)
+
+    def _get_prop_ancestors(self):
+        """Get relevant ancestor(s) for attribute propagation
+
+        Returns
+        -------
+        list
+            List of ancestors.
+        """
+        if isinstance(self._propagate_how, int):
+            return self._ancestors[self._propagate_how:self._propagate_how+1]
+        if self._propagate_how == 'first':
+            return self._ancestors
+        if self._propagate_how == 'last':
+            return self._ancestors[::-1]
+        raise ValueError("propagate_how has to be an index, 'first', or "
+                         "'last'.")
 
     def _get(self, key):
         # We need to copy here: else any _proc_func that acts inplace would
         # change the ancestor value.
-        return self._proc_func(copy(self._ancestor[key]))
+        return self._proc_func(*(copy(a[key]) for a in self._ancestors))
 
     def __repr__(self):
-        msg = "{0} processed through {1}. Original repr:\n".format(
-                type(self._ancestor).__name__, self._proc_func.__name__)
-        old = '\n'.join("    " + ln for ln in repr(self._ancestor).split('\n'))
-        return msg + old
+        anc_str = ", ".join(type(a).__name__ for a in self._ancestors)
+        msg = '({0},) processed through {1}. Original repr:\n    '.format(
+            anc_str, self._proc_func.__name__)
+        old = [repr(a).replace('\n', '\n    ') for a in self._ancestors]
+        return msg + "\n    ----\n    ".join(old)
 
     def __len__(self):
-        return self._ancestor.__len__()
+        return min(a.__len__() for a in self._ancestors)
```

Comment:
Why not `len(a)`?

Reply:
I guess because it was `self._ancestor.__len__()` orginally. Since I did not know why, I did not change it.